File: sass-api/internal/services/menu_service.go
package services

import (
	"errors"
	"sort"

	"github.com/Aebroyx/sass-api/internal/config"
	"github.com/Aebroyx/sass-api/internal/domain/models"
	"github.com/Aebroyx/sass-api/internal/pagination"
	"gorm.io/gorm"
)

type MenuService struct {
	db     *gorm.DB
	config *config.Config
}

func NewMenuService(db *gorm.DB, config *config.Config) *MenuService {
	return &MenuService{
		db:     db,
		config: config,
	}
}

// GetAllMenus retrieves all menus with pagination
func (s *MenuService) GetAllMenus(params pagination.QueryParams) (*pagination.PaginatedResponse, error) {
	config := pagination.PaginationConfig{
		Model:         &models.Menu{},
		BaseCondition: map[string]interface{}{},
		SearchFields:  []string{"name", "path"},
		FilterFields: map[string]string{
			"name":      "name",
			"path":      "path",
			"is_active": "is_active",
			"parent_id": "parent_id",
		},
		SortFields:   []string{"name", "order_index", "created_at"},
		DefaultSort:  "order_index",
		DefaultOrder: "ASC",
	}

	paginator := pagination.NewPaginator(s.db)
	return paginator.Paginate(params, config)
}

// GetMenuByID retrieves a menu by ID
func (s *MenuService) GetMenuByID(id uint) (*models.Menu, error) {
	var menu models.Menu
	if err := s.db.Preload("Children").First(&menu, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("menu not found")
		}
		return nil, err
	}
	return &menu, nil
}

// GetMenuTree retrieves all menus as a tree structure
func (s *MenuService) GetMenuTree() ([]models.MenuResponse, error) {
	var menus []models.Menu
	if err := s.db.Where("is_active = ?", true).Order("order_index ASC").Find(&menus).Error; err != nil {
		return nil, err
	}

	return buildMenuTree(menus, nil), nil
}

// buildMenuTree builds a tree structure from flat menu list
func buildMenuTree(menus []models.Menu, parentID *uint) []models.MenuResponse {
	var result []models.MenuResponse

	for _, menu := range menus {
		// Check if this menu belongs to the current parent
		if (parentID == nil && menu.ParentID == nil) || (parentID != nil && menu.ParentID != nil && *parentID == *menu.ParentID) {
			menuResp := models.MenuResponse{
				ID:         menu.ID,
				Name:       menu.Name,
				Path:       menu.Path,
				Icon:       menu.Icon,
				OrderIndex: menu.OrderIndex,
				ParentID:   menu.ParentID,
				IsActive:   menu.IsActive,
				CreatedAt:  menu.CreatedAt,
				UpdatedAt:  menu.UpdatedAt,
				Children:   buildMenuTree(menus, &menu.ID),
			}
			result = append(result, menuResp)
		}
	}

	return result
}

// GetUserMenus retrieves menus accessible by a user with effective permissions
func (s *MenuService) GetUserMenus(userID uint, roleID uint) ([]models.MenuWithPermissions, error) {
	// Get menus from role
	var roleMenus []models.RoleMenu
	if err := s.db.Preload("Menu").Where("role_id = ?", roleID).Find(&roleMenus).Error; err != nil {
		return nil, err
	}

	// Get direct user menus
	var userMenus []models.UserMenu
	if err := s.db.Preload("Menu").Where("user_id = ?", userID).Find(&userMenus).Error; err != nil {
		return nil, err
	}

	// Get user permission overrides
	var rightsAccess []models.RightsAccess
	if err := s.db.Where("user_id = ?", userID).Find(&rightsAccess).Error; err != nil {
		return nil, err
	}

	// Create map for quick lookup of overrides
	overrideMap := make(map[uint]models.RightsAccess)
	for _, ra := range rightsAccess {
		overrideMap[ra.MenuID] = ra
	}

	// Build menu map with permissions
	menuMap := make(map[uint]models.MenuWithPermissions)

	// Add menus from role with default permissions
	for _, rm := range roleMenus {
		if !rm.Menu.IsActive {
			continue
		}

		permissions := models.EffectivePermissions{
			CanRead:   rm.CanRead,
			CanWrite:  rm.CanWrite,
			CanUpdate: rm.CanUpdate,
			CanDelete: rm.CanDelete,
		}

		// Apply user overrides if exist
		if override, ok := overrideMap[rm.MenuID]; ok {
			if override.CanRead != nil {
				permissions.CanRead = *override.CanRead
			}
			if override.CanWrite != nil {
				permissions.CanWrite = *override.CanWrite
			}
			if override.CanUpdate != nil {
				permissions.CanUpdate = *override.CanUpdate
			}
			if override.CanDelete != nil {
				permissions.CanDelete = *override.CanDelete
			}
		}

		menuMap[rm.MenuID] = models.MenuWithPermissions{
			ID:          rm.Menu.ID,
			Name:        rm.Menu.Name,
			Path:        rm.Menu.Path,
			Icon:        rm.Menu.Icon,
			OrderIndex:  rm.Menu.OrderIndex,
			ParentID:    rm.Menu.ParentID,
			IsActive:    rm.Menu.IsActive,
			Permissions: permissions,
		}
	}

	// Add direct user menus (with default read permission if not already in map)
	for _, um := range userMenus {
		if !um.Menu.IsActive {
			continue
		}

		if _, exists := menuMap[um.MenuID]; !exists {
			permissions := models.EffectivePermissions{
				CanRead:   true, // Default for direct assignment
				CanWrite:  false,
				CanUpdate: false,
				CanDelete: false,
			}

			// Apply user overrides if exist
			if override, ok := overrideMap[um.MenuID]; ok {
				if override.CanRead != nil {
					permissions.CanRead = *override.CanRead
				}
				if override.CanWrite != nil {
					permissions.CanWrite = *override.CanWrite
				}
				if override.CanUpdate != nil {
					permissions.CanUpdate = *override.CanUpdate
				}
				if override.CanDelete != nil {
					permissions.CanDelete = *override.CanDelete
				}
			}

			menuMap[um.MenuID] = models.MenuWithPermissions{
				ID:          um.Menu.ID,
				Name:        um.Menu.Name,
				Path:        um.Menu.Path,
				Icon:        um.Menu.Icon,
				OrderIndex:  um.Menu.OrderIndex,
				ParentID:    um.Menu.ParentID,
				IsActive:    um.Menu.IsActive,
				Permissions: permissions,
			}
		}
	}

	// Convert map to slice
	result := make([]models.MenuWithPermissions, 0, len(menuMap))
	for _, menu := range menuMap {
		result = append(result, menu)
	}

	// Sort by order_index
	sort.Slice(result, func(i, j int) bool {
		return result[i].OrderIndex < result[j].OrderIndex
	})

	// Build tree structure
	return buildMenuWithPermissionsTree(result, nil), nil
}

// buildMenuWithPermissionsTree builds a tree structure from flat menu list with permissions
func buildMenuWithPermissionsTree(menus []models.MenuWithPermissions, parentID *uint) []models.MenuWithPermissions {
	var result []models.MenuWithPermissions

	for _, menu := range menus {
		if (parentID == nil && menu.ParentID == nil) || (parentID != nil && menu.ParentID != nil && *parentID == *menu.ParentID) {
			menu.Children = buildMenuWithPermissionsTree(menus, &menu.ID)
			result = append(result, menu)
		}
	}

	return result
}

// CreateMenu creates a new menu
func (s *MenuService) CreateMenu(req *models.CreateMenuRequest) (*models.MenuResponse, error) {
	// Validate parent exists if provided
	if req.ParentID != nil {
		var parent models.Menu
		if err := s.db.First(&parent, *req.ParentID).Error; err != nil {
			return nil, errors.New("parent menu not found")
		}
	}

	menu := models.Menu{
		Name:       req.Name,
		Path:       req.Path,
		Icon:       req.Icon,
		OrderIndex: req.OrderIndex,
		ParentID:   req.ParentID,
		IsActive:   req.IsActive,
	}

	if err := s.db.Create(&menu).Error; err != nil {
		return nil, err
	}

	return &models.MenuResponse{
		ID:         menu.ID,
		Name:       menu.Name,
		Path:       menu.Path,
		Icon:       menu.Icon,
		OrderIndex: menu.OrderIndex,
		ParentID:   menu.ParentID,
		IsActive:   menu.IsActive,
		CreatedAt:  menu.CreatedAt,
		UpdatedAt:  menu.UpdatedAt,
	}, nil
}

// UpdateMenu updates an existing menu
func (s *MenuService) UpdateMenu(id uint, req *models.UpdateMenuRequest) (*models.MenuResponse, error) {
	var menu models.Menu
	if err := s.db.First(&menu, id).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("menu not found")
		}
		return nil, err
	}

	// Validate parent exists if provided and prevent circular reference
	if req.ParentID != nil {
		if *req.ParentID == id {
			return nil, errors.New("menu cannot be its own parent")
		}
		var parent models.Menu
		if err := s.db.First(&parent, *req.ParentID).Error; err != nil {
			return nil, errors.New("parent menu not found")
		}
	}

	// Update fields
	menu.Name = req.Name
	menu.Path = req.Path
	menu.Icon = req.Icon
	menu.OrderIndex = req.OrderIndex
	menu.ParentID = req.ParentID
	menu.IsActive = req.IsActive

	if err := s.db.Save(&menu).Error; err != nil {
		return nil, err
	}

	return &models.MenuResponse{
		ID:         menu.ID,
		Name:       menu.Name,
		Path:       menu.Path,
		Icon:       menu.Icon,
		OrderIndex: menu.OrderIndex,
		ParentID:   menu.ParentID,
		IsActive:   menu.IsActive,
		CreatedAt:  menu.CreatedAt,
		UpdatedAt:  menu.UpdatedAt,
	}, nil
}

// DeleteMenu deletes a menu (soft delete)
func (s *MenuService) DeleteMenu(id uint) error {
	var menu models.Menu
	if err := s.db.First(&menu, id).Error; err != nil {
		return errors.New("menu not found")
	}

	// Check if menu has children
	var childCount int64
	s.db.Model(&models.Menu{}).Where("parent_id = ?", id).Count(&childCount)
	if childCount > 0 {
		return errors.New("cannot delete menu with children")
	}

	// Delete related role_menus and user_menus
	s.db.Where("menu_id = ?", id).Delete(&models.RoleMenu{})
	s.db.Where("menu_id = ?", id).Delete(&models.UserMenu{})
	s.db.Where("menu_id = ?", id).Delete(&models.RightsAccess{})

	return s.db.Delete(&menu).Error
}

File: sass-ui/src/services/menuService.ts
import axiosInstance, { handleApiError } from '@/lib/axios';

// Types
export interface MenuResponse {
  id: number;
  name: string;
  path: string;
  icon: string;
  order_index: number;
  parent_id: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  children?: MenuResponse[];
}

export interface RoleMenuResponse {
  id: number;
  role_id: number;
  menu_id: number;
  can_read: boolean;
  can_write: boolean;
  can_update: boolean;
  can_delete: boolean;
  menu: MenuResponse;
  created_at: string;
  updated_at: string;
}

export interface RightsAccessResponse {
  id: number;
  user_id: number;
  menu_id: number;
  can_read: boolean | null;
  can_write: boolean | null;
  can_update: boolean | null;
  can_delete: boolean | null;
  menu: MenuResponse;
  created_at: string;
  updated_at: string;
}

export interface EffectivePermissions {
  can_read: boolean;
  can_write: boolean;
  can_update: boolean;
  can_delete: boolean;
}

export interface MenuWithPermissions {
  id: number;
  name: string;
  path: string;
  icon: string;
  order_index: number;
  parent_id: number | null;
  is_active: boolean;
  children?: MenuWithPermissions[];
  permissions: EffectivePermissions;
}

export interface UserMenuPermission {
  menu_id: number;
  can_read: boolean | null;
  can_write: boolean | null;
  can_update: boolean | null;
  can_delete: boolean | null;
}

export interface BulkUserRightsAccessRequest {
  permissions: UserMenuPermission[];
}

export interface CreateMenuRequest {
  name: string;
  path: string;
  icon: string;
  order_index: number;
  parent_id: number | null;
  is_active: boolean;
}

export interface UpdateMenuRequest {
  name: string;
  path: string;
  icon: string;
  order_index: number;
  parent_id: number | null;
  is_active: boolean;
}

export interface GetAllMenusParams {
  page: number;
  pageSize: number;
  search?: string;
  sortBy?: string;
  sortDesc?: boolean;
  filters?: Record<string, string>;
}

export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

interface ApiResponse<T> {
  status: string;
  message: string;
  data: T;
}

class MenuService {
  // Get current user's accessible menus with permissions
  async getUserMenus(): Promise<MenuWithPermissions[]> {
    try {
      const response = await axiosInstance.get<ApiResponse<MenuWithPermissions[]>>('/menus/user');
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Get all menus as a tree structure
  async getMenuTree(): Promise<MenuResponse[]> {
    try {
      const response = await axiosInstance.get<ApiResponse<MenuResponse[]>>('/menus/tree');
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Get menus assigned to a role with permissions
  async getRoleMenus(roleId: number): Promise<RoleMenuResponse[]> {
    try {
      const response = await axiosInstance.get<ApiResponse<RoleMenuResponse[]>>(`/role/${roleId}/menus`);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Get user's rights access (permission overrides)
  async getUserRightsAccess(userId: number): Promise<RightsAccessResponse[]> {
    try {
      const response = await axiosInstance.get<ApiResponse<RightsAccessResponse[]>>(`/rights-access/user/${userId}`);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Bulk save user's rights access
  async bulkSaveUserRightsAccess(userId: number, permissions: UserMenuPermission[]): Promise<RightsAccessResponse[]> {
    try {
      const response = await axiosInstance.post<ApiResponse<RightsAccessResponse[]>>(
        `/rights-access/user/${userId}/bulk`,
        { permissions }
      );
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Delete all user's rights access
  async deleteAllUserRightsAccess(userId: number): Promise<void> {
    try {
      await axiosInstance.delete(`/rights-access/user/${userId}`);
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Get all menus with pagination
  async getAllMenus(params: GetAllMenusParams): Promise<PaginatedResponse<MenuResponse>> {
    try {
      const response = await axiosInstance.get<ApiResponse<PaginatedResponse<MenuResponse>>>('/menus', {
        params,
      });
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Get menu by ID
  async getMenuById(id: string | number): Promise<MenuResponse> {
    try {
      const response = await axiosInstance.get<ApiResponse<MenuResponse>>(`/menu/${id}`);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Create menu
  async createMenu(data: CreateMenuRequest): Promise<MenuResponse> {
    try {
      const response = await axiosInstance.post<ApiResponse<MenuResponse>>('/menu/create', data);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Update menu
  async updateMenu(id: string | number, data: UpdateMenuRequest): Promise<MenuResponse> {
    try {
      const response = await axiosInstance.put<ApiResponse<MenuResponse>>(`/menu/${id}`, data);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  // Delete menu
  async deleteMenu(id: string | number): Promise<void> {
    try {
      await axiosInstance.delete(`/menu/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  }
}

export const menuService = new MenuService();
